.github/workflows/ci.yml

```diff
@@ -241,76 +241,106 @@ jobs:
             gcr.io/iree-oss/swiftshader-bleeding-edge@sha256:89656f131eb6efc97585f6b71b2cedbaa9ae79b539362d2545b19a9bbaf7501c \
             ./build_tools/bazel/build_core.sh
 
-  test_all:
+  test:
     needs: [setup, build_all]
     if: fromJson(needs.setup.outputs.should-run)
-    runs-on:
-      - self-hosted # must come first
-      - runner-group=${{ needs.setup.outputs.runner-group }}
-      - environment=${{ needs.setup.outputs.runner-env }}
-      - cpu
-      - os-family=Linux
     env:
       BUILD_DIR: ${{ needs.build_all.outputs.build-dir }}
       BUILD_DIR_ARCHIVE: ${{ needs.build_all.outputs.build-dir-archive }}
       BUILD_DIR_GCS_ARTIFACT: ${{ needs.build_all.outputs.build-dir-gcs-artifact }}
-    steps:
-      - name: "Checking out repository"
-        uses: actions/checkout@ac593985615ec2ede58e132d2e21d2b1cbd6127c # v3.3.0
-        with:
-          submodules: true
-      - name: "Downloading build dir archive"
-        run: gcloud storage cp "${BUILD_DIR_GCS_ARTIFACT}" "${BUILD_DIR_ARCHIVE}"
-      - name: "Extracting build dir archive"
-        run: tar -xf "${BUILD_DIR_ARCHIVE}"
-      - name: "Testing all"
-        run: |
-          ./build_tools/github_actions/docker_run.sh \
-            --env IREE_CUDA_DISABLE=1 \
-            gcr.io/iree-oss/swiftshader@sha256:beeb2c9853bb41375f5c9ccfb19b1ed35655c4354cc71fbcf124fe9103a7f543 \
-            ./build_tools/cmake/ctest_all.sh \
-            "${BUILD_DIR}"
-
-  test_gpu:
-    needs: [setup, build_all]
-    if: fromJson(needs.setup.outputs.should-run)
+      DOCKER_ENV_FILE: docker.env
+      DOCKER_ARG_FILE: docker-args.txt
+      EXEC_FILE: run.sh
+      # We could enumerate all the differences in the matrix spec, but for now
+      # there's basically just one split between CPU and GPU and there are two
+      # GPU jobs, so it's annoying to write everything twice. We insted
+      # construct some stuff dynamically in a step that's keyed off of is-gpu.
+      CPU_IMAGE: gcr.io/iree-oss/swiftshader@sha256:beeb2c9853bb41375f5c9ccfb19b1ed35655c4354cc71fbcf124fe9103a7f543
+      GPU_IMAGE: gcr.io/iree-oss/nvidia@sha256:1717431fd46b8b1e96d95fa72508e3e3eacb5c95f1245b9b7dbeec23ae823d02
+    strategy:
```

Comment:
If we choose to use matrix, maybe add `continue-on-error: true` to not cancel other jobs when one of them is failing (https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs#handling-failures)

Reply:
Yeah or `fail-fast: false`. I think we could consider cancelling the a100 job if the regular gpu job fails, but maybe that would more frequently be because of a flake than something useful.